collecting with brackets is deprecated

gustafsson opened the pull request:
Resolves `WARNING: [a] concatenation is deprecated; use collect(a) instead`

nalimilan commented:
Thanks. The tests failed on Julia nightlies because the name of the column now depends on that of lambda function (which cannot be predicted reliably). This can be fixed by indexing the data frames with `[1]` before comparing them. But are you sure you need a new test? This line seems to be covered already (deprecations do not cause failures): https://coveralls.io/builds/5604859/source?filename=src%2Fgroupeddataframe%2Fgrouping.jl#L248

Also, I wonder whether we actually need to call `collect` here. Some functions could return e.g. generators, and it could be useful. I don't see why we should impose returning an array.

@tshort Opinions?

tshort commented:
I'm not where I can test anything, but I think you will need `collect`. If `f` is a reduction function, `collect` will convert it to an array. If `f` is an element-wise function, it will keep the result as an array.

nalimilan commented:
@tshort Yes, but do we really need an array? Since the result is an `Array{Any}`, having some elements be scalars (or even ranges or generators) isn't an issue.

gustafsson commented:
I replaced the lambda function with a regular element-wise function in the added commit above.

I think all other tests of `colwise` uses reduction functions because I don't see this warning without the test I added.

The elements of `colwise` should be of type `AbstractVector` according to [`test/grouping.jl`](https://github.com/JuliaStats/DataFrames.jl/blob/master/test/grouping.jl#L20). Perhaps this could be relaxed?

Would it make sense to use something like this instead of `collect`? (it doesn't make a difference for generators but it would save an unnecessary copy of arrays)
```
makeabstractvector(x::AbstractVector) = x
makeabstractvector(x) = collect(x)
```

---

Without using `collect` I get these errors (julia homebrew `Version 0.4.5 (2016-03-18 00:58 UTC)`):

```
$ cd ~/.julia/v0.4/DataFrames/test
$ julia runtests.jl
```
...
```
	FAILED: data.jl
LoadError: MethodError: `_isnullable` has no method matching _isnullable(::Int64)
 [inlined code] from ~/.julia/v0.4/DataFrames/test/runtests.jl:44
 in anonymous at no file:0
 in include at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
 in include_from_node1 at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
 in process_options at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
 in _start at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
while loading ~/.julia/v0.4/DataFrames/test/data.jl, in expression starting on line 107
```
...
```
	FAILED: grouping.jl
LoadError: test failed: all((T->begin  # ~/.julia/v0.4/DataFrames/test/grouping.jl, line 20:
            T <: AbstractVector
        end),map(typeof,colwise([sum],df)))
 in expression: all((T->begin  # ~/.julia/v0.4/DataFrames/test/grouping.jl, line 20:
            T <: AbstractVector
        end),map(typeof,colwise([sum],df)))
 [inlined code] from ~/.julia/v0.4/DataFrames/test/runtests.jl:44
 in anonymous at no file:0
 in include at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
 in include_from_node1 at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
 in process_options at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
 in _start at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
while loading ~/.julia/v0.4/DataFrames/test/grouping.jl, in expression starting on line 20
```

gustafsson commented:
@nalimilan `aggregate(DataFrame(a=1),abs)` still gives the same warning on master. It could be worth nothing that in the documentation for aggregate it says:

> Each `fs` should return a value or vector. All returns must be the same length.

The problem addressed by this PR seems to only arise with functions that returns vectors (such as `abs`). So for instance `aggregate(DataFrame(a=1),sum)` isn't a problem in master.

nalimilan commented:
Let's go with this as it fixes the warning. We can always relax the requirements later, but that's kind of a separate issue. sorry it took so long (do not hesitate to bump from time to time).

nalimilan commented:
Woops, this actually made the tests fail after the port the `Nullable`. Could you check that https://github.com/JuliaStats/DataFrames.jl/pull/1075 is correct?
